File: py/kintsugi/tools.py
"""Read-only tools the model proposer may call.

The model never executes code and never touches live objects. It can only
ask the harness to read files, grep, and list importers, and the answers
come back as bounded text in the prompt. This is the safe half of NOOA's
pass-by-reference (see docs/NOOA.md): the *effect* of context navigation
through declared, inspectable tools — not the mechanism of letting the
model run code against the repo.

Every result is capped so the prompt cannot balloon; a tool that reads the
whole repo is a tool that defeats the point of a bounded context. Paths are
resolved inside the source root and refused if they escape it.

Mirrors `src/tools.ts` in the TypeScript engine.
"""

import glob
import os
import re

# Result caps, mirrored from src/tools.ts.
MAX_READ_LINES = 400
MAX_READ_CHARS = 20_000
MAX_GREP_MATCHES = 40
MAX_IMPORTERS = 20

_SKIP = re.compile(r"(^|[\\/])(node_modules|dist|build|\.git|\.venv)([\\/]|$)")
_GREP_EXTS = (".py", ".ts", ".tsx", ".js", ".jsx", ".mjs", ".cjs", ".go", ".rs", ".md")


class ToolRunner:
    """Executes one declared tool call against the repo, returning text.

    `graph` is the import graph the loop already built (blast radius); when
    absent, the `importers` tool reports that it is unavailable rather than
    guessing.
    """

    def __init__(self, source_root: str, graph: dict | None = None):
        self.source_root = os.path.abspath(source_root)
        self.graph = graph

    def run(self, name: str, args: dict) -> str:
        args = args or {}
        if name == "read_file":
            return self._read_file(args)
        if name == "grep":
            return self._grep(args)
        if name == "importers":
            return self._importers(args)
        return f"error: unknown tool {name!r} — use one of: read_file, grep, importers"

    def _resolve(self, path) -> str | None:
        """Resolve a model-supplied relative path inside the source root."""
        if not isinstance(path, str) or not path:
            return None
        abs_path = os.path.abspath(os.path.join(self.source_root, path))
        rel = os.path.relpath(abs_path, self.source_root).replace("\\", "/")
        if rel.startswith("..") or rel == "..":
            return None
        return abs_path

    def _read_file(self, args: dict) -> str:
        abs_path = self._resolve(args.get("path"))
        if abs_path is None:
            return "error: path must be a relative path inside the source root"
        if not os.path.isfile(abs_path):
            return f"error: no such file: {args.get('path')}"
        try:
            with open(abs_path, "r", encoding="utf-8", errors="replace", newline="") as fh:
                lines = fh.read().split("\n")
        except OSError as err:
            return f"error: cannot read {args.get('path')}: {err}"
        try:
            start = max(int(args.get("start") or 1), 1)
        except (TypeError, ValueError):
            start = 1
        end = min(len(lines), start + MAX_READ_LINES - 1)
        try:
            requested_end = int(args.get("end") or 0)
            if requested_end >= start:
                end = min(end, requested_end)
        except (TypeError, ValueError):
            pass
        rel = os.path.relpath(abs_path, self.source_root).replace("\\", "/")
        out = [f"{rel} (lines {start}-{end} of {len(lines)}):"]
        for i in range(start - 1, end):
            out.append(f"{i + 1}\t{lines[i][:200]}")
        return "\n".join(out)[:MAX_READ_CHARS]

    def _grep(self, args: dict) -> str:
        pattern = args.get("pattern")
        if not isinstance(pattern, str) or not pattern:
            return "error: grep needs a 'pattern' string"
        try:
            re_ = re.compile(pattern)
        except re.error as err:
            return f"error: invalid regex: {err}"
        base = self.source_root
        if args.get("path"):
            abs_path = self._resolve(args.get("path"))
            if abs_path is None:
                return "error: path must be a relative path inside the source root"
            if not os.path.exists(abs_path):
                return f"error: no such path: {args.get('path')}"
            base = abs_path
        files = ([base] if os.path.isfile(base)
                 else [f for f in glob.glob(os.path.join(base, "**", "*"), recursive=True)
                       if os.path.isfile(f)])
        matches = []
        for f in files:
            rel = os.path.relpath(f, self.source_root).replace("\\", "/")
            if _SKIP.search(rel) or not f.endswith(_GREP_EXTS):
                continue
            try:
                with open(f, "r", encoding="utf-8", errors="replace", newline="") as fh:
                    lines = fh.read().split("\n")
            except OSError:
                continue
            for i, ln in enumerate(lines, 1):
                if re_.search(ln):
                    matches.append(f"{rel}:{i}: {ln[:140]}")
                    if len(matches) >= MAX_GREP_MATCHES:
                        return "\n".join(matches) + f"\n… {len(matches)} matches shown, more available"
        if not matches:
            return f"no matches for {pattern!r}"
        return "\n".join(matches)

    def _importers(self, args: dict) -> str:
        if not self.graph:
            return "error: import graph is not available for this run"
        abs_path = self._resolve(args.get("path"))
        if abs_path is None:
            return "error: path must be a relative path inside the source root"
        imp = self.graph["importers"].get(abs_path)
        if not imp:
            return f"no module imports {args.get('path')}"
        rels = sorted(
            os.path.relpath(p, self.source_root).replace("\\", "/") for p in imp
        )[:MAX_IMPORTERS]
        head = f"{len(imp)} module(s) import {args.get('path')}:"
        return head + "\n" + "\n".join(f"- {r}" for r in rels)

File: src/tools.ts
import { readFileSync, globSync, existsSync, statSync } from 'node:fs';
import { resolve, relative, isAbsolute } from 'node:path';
import type { ImportGraph } from './imports.js';

/**
 * Read-only tools the model proposer may call.
 *
 * The model never executes code and never touches live objects. It can only
 * ask the harness to read files, grep, and list importers, and the answers
 * come back as bounded text in the prompt. This is the safe half of NOOA's
 * pass-by-reference (see docs/NOOA.md): the *effect* of context navigation
 * through declared, inspectable tools — not the mechanism of letting the
 * model run code against the repo.
 *
 * Every result is capped so the prompt cannot balloon; a tool that reads the
 * whole repo is a tool that defeats the point of a bounded context. Paths
 * are resolved inside the source root and refused if they escape it.
 *
 * Mirrors `py/kintsugi/tools.py` in the Python engine.
 */

// Result caps, mirrored from py/kintsugi/tools.py.
export const MAX_READ_LINES = 400;
export const MAX_READ_CHARS = 20_000;
export const MAX_GREP_MATCHES = 40;
export const MAX_IMPORTERS = 20;

const SKIP = /(^|[\\/])(node_modules|dist|build|\.git|\.venv)([\\/]|$)/;
const GREP_EXTS = /\.(py|ts|tsx|js|jsx|mjs|cjs|go|rs|md)$/;

export class ToolRunner {
  private sourceRoot: string;
  private graph?: ImportGraph;

  constructor(sourceRoot: string, graph?: ImportGraph) {
    this.sourceRoot = resolve(sourceRoot);
    this.graph = graph;
  }

  /** Execute one declared tool call, returning bounded text. */
  run(name: string, args: Record<string, unknown>): string {
    switch (name) {
      case 'read_file': return this.readFile(args);
      case 'grep': return this.grep(args);
      case 'importers': return this.importers(args);
      default:
        return `error: unknown tool ${JSON.stringify(name)} — use one of: read_file, grep, importers`;
    }
  }

  /** Resolve a model-supplied relative path inside the source root. */
  private resolvePath(path: unknown): string | null {
    if (typeof path !== 'string' || !path) return null;
    const abs = resolve(this.sourceRoot, path);
    const rel = relative(this.sourceRoot, abs);
    // `isAbsolute(rel)` catches cross-drive escapes on Windows, where
    // relative() yields an absolute path instead of a `..` chain.
    if (rel.startsWith('..') || rel === '..' || isAbsolute(rel)) return null;
    return abs;
  }

  private readFile(args: Record<string, unknown>): string {
    const path = args.path;
    const abs = this.resolvePath(path);
    if (!abs) return 'error: path must be a relative path inside the source root';
    if (!existsSync(abs) || !statSync(abs).isFile()) {
      return `error: no such file: ${String(path)}`;
    }
    let text: string;
    try {
      text = readFileSync(abs, 'utf8');
    } catch (err) {
      return `error: cannot read ${String(path)}: ${(err as Error).message}`;
    }
    const lines = text.split('\n');
    let start = 1;
    const s = Number(args.start);
    if (Number.isFinite(s)) start = Math.max(s, 1);
    let end = Math.min(lines.length, start + MAX_READ_LINES - 1);
    const e = Number(args.end);
    if (Number.isFinite(e) && e >= start) end = Math.min(end, e);
    const rel = relative(this.sourceRoot, abs).replace(/\\/g, '/');
    const out = [`${rel} (lines ${start}-${end} of ${lines.length}):`];
    for (let i = start - 1; i < end; i++) out.push(`${i + 1}\t${lines[i].slice(0, 200)}`);
    return out.join('\n').slice(0, MAX_READ_CHARS);
  }

  private grep(args: Record<string, unknown>): string {
    const pattern = args.pattern;
    if (typeof pattern !== 'string' || !pattern) return "error: grep needs a 'pattern' string";
    let re: RegExp;
    try {
      re = new RegExp(pattern);
    } catch (err) {
      return `error: invalid regex: ${(err as Error).message}`;
    }
    let base = this.sourceRoot;
    if (args.path) {
      const abs = this.resolvePath(args.path);
      if (!abs) return 'error: path must be a relative path inside the source root';
      if (!existsSync(abs)) return `error: no such path: ${String(args.path)}`;
      base = abs;
    }
    const files = statSync(base).isFile()
      ? [base]
      : globSync('**/*', { cwd: base }).map((f) => resolve(base, f)).filter((f) => statSync(f).isFile());
    const matches: string[] = [];
    for (const f of files) {
      const rel = relative(this.sourceRoot, f).replace(/\\/g, '/');
      if (SKIP.test(rel) || !GREP_EXTS.test(rel)) continue;
      let content: string;
      try {
        content = readFileSync(f, 'utf8');
      } catch {
        continue;
      }
      const lines = content.split('\n');
      for (let i = 0; i < lines.length; i++) {
        if (re.test(lines[i])) {
          matches.push(`${rel}:${i + 1}: ${lines[i].slice(0, 140)}`);
          if (matches.length >= MAX_GREP_MATCHES) {
            return `${matches.join('\n')}\n… ${matches.length} matches shown, more available`;
          }
        }
      }
    }
    return matches.length ? matches.join('\n') : `no matches for ${JSON.stringify(pattern)}`;
  }

  private importers(args: Record<string, unknown>): string {
    if (!this.graph) return 'error: import graph is not available for this run';
    const path = args.path;
    const abs = this.resolvePath(path);
    if (!abs) return 'error: path must be a relative path inside the source root';
    const imp = this.graph.importers.get(abs);
    if (!imp || imp.size === 0) return `no module imports ${String(path)}`;
    const rels = [...imp]
      .map((p) => relative(this.sourceRoot, p).replace(/\\/g, '/'))
      .sort()
      .slice(0, MAX_IMPORTERS);
    return `${imp.size} module(s) import ${String(path)}:\n` + rels.map((r) => `- ${r}`).join('\n');
  }
}
